### PR TITLE
tests: download core and ubuntu-core at most once

### PR DESCRIPTION
**Pull request by zyga**
This patch relies on snapd's ability to resume partial downloads.  In
the one-off preparation phase we download a set of snaps, chop off the
last byte and move them to /var/lib/snapd/snaps adding the .partial
suffix to enable the download beavior. Each test that needs to pull in
one of those two fill then take advantage of the existing file and will
do minimal network traffic required to achieve this.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: tests/lib/prepare.sh [hunk 1]
```diff
@@ -152,6 +152,26 @@ EOF
...
+            cd /tmp
+            for snap_name in ${PRE_CACHE_SNAPS:-}; do
+                snap download $snap_name
+            done
+            for snap_file in *.snap; do
+                mv $snap_file $snap_file.partial
+                # There is a bug in snapd where partial file must be a proper
+                # prefix of the full file or we make a wrong request to the
+                # store.
+                truncate --size=$(expr $(stat $snap_file.partial -c %s) - 1) $snap_file.partial
```
Comment (chipaca) [inline tests/lib/prepare.sh]: `truncate --size=-1` would DWYW :smiley:

Comment (zyga) [inline tests/lib/prepare.sh]: Oh, I didn't know that works. Nice!

**Comment by zyga** [comment]
Merging, I just saw another test failing on "whooping" 100KB/s download speed that ended up with timeout.

File: tests/lib/prepare.sh [hunk 2]
```diff
@@ -152,6 +152,26 @@ EOF
...
+            cd /tmp
+            for snap_name in ${PRE_CACHE_SNAPS:-}; do
+                snap download $snap_name
+            done
+            for snap_file in *.snap; do
+                mv $snap_file $snap_file.partial
+                # There is a bug in snapd where partial file must be a proper
+                # prefix of the full file or we make a wrong request to the
+                # store.
+                truncate --size=-1 $snap_file.partial
```
Comment (chipaca) [inline tests/lib/prepare.sh]: @zyga did you land a patch to drop this `truncate`?

**Pull request merged**